src/meep.hpp

```diff
@@ -1463,27 +1474,31 @@ class fields {
   // cw_fields.cpp:
   bool solve_cw(double tol, int maxiters, std::complex<double> frequency, int L=2);
   bool solve_cw(double tol = 1e-8, int maxiters = 10000, int L=2);
-
+#
   // sources.cpp:
   double last_source_time();
   void add_point_source(component c, double freq, double width, double peaktime,
                         double cutoff, const vec &, std::complex<double> amp = 1.0,
                         int is_continuous = 0);
   void add_point_source(component c, const src_time &src,
                         const vec &, std::complex<double> amp = 1.0);
+
+  // if indicator is nonzero, sources are added only at points for which
+  // indicator->in_source(p) returns true.
   void add_volume_source(component c, const src_time &src, const volume &where_,
                          std::complex<double> *arr, size_t dim1, size_t dim2, size_t dim3,
-                         std::complex<double> amp);
+                         std::complex<double> amp, source_indicator *indicator=0);
```

Comment:
This implementation dates back to before the merge of `meepgeom` into `meep`, at which point `libmeep` had no knowledge of `geometric_objects`, so any implementation of this feature necessarily involved some sort of external function access. The feature could have been implemented using an amplitude function in python, but I was concerned that would be slow. Now that `geometric_objects` are available in libmeep, I'm sure you could implement the feature in a way that routed everything through the existing `complex<double> A(const vec &)` interface, perhaps with the `object` in question stored as a global variable, if you thought that would be a significant improvement over the above in terms of code efficiency, readability, or some other metric.

Reply:
It's not a question of efficiency, but I'd rather not have two arguments (`indicator` and `A`) that are completely redundant in the underlying C++ code.

Reply:
Three of the four prototypes for `add_volume_source` actually *don't* take the `A` argument, so `indicator` is not redundant in those cases.

Even in the one remaining case, although it is true in principle that the `geometric_object` functionality could be routed through the existing `A` interface, I'm not seeing how this could be done gracefully. In this case the caller has provided an `A` function plus a separate `geometric_object` that don't know about each other, and we would have to write our own new `A` function that **(1)** tests the point for inclusion in the object, and if that passes **(2)** separately call the user's original `A` function to get the amplitude. Of course the `A` interface doesn't allow for passage of any user data, so both the geometric object and the caller's original `A` function would need to be passed to our wrapper `A` function via some other mechanism, probably global variables. I agree that it would be *possible*, but it feels sort of like bending over backwards to use the legacy `A` interface in ways for which it wasn't intended.

Plus, any way you slice it the prototype for `add_volume_source` is going to have to be modified to accept at least one new parameter for the `geometric_object` in one form or another, so it's not as if the legacy interface could simply be preserved in full pristine form anyway. My solution adds precisely one new parameter, almost no additional code, and zero global variables. Again, I agree that it would be possible to do what I think you have in mind, but it doesn't seem an obvious improvement over my solution---which has the further virtue of being already implemented and ready for users to start exploring.

Reply:
Even if you have a new high-level interfact that takes a geometric_object, internally it can just construct an amplitude function and pass that to the lower-level `add_volume_source` routine.  That way you don't need the whole `meep::source_indicator` thing, and low-level functions like `src_vol_chunkloop` etcetera don't need to be modified at all.